logic.py

```diff
@@ -81,9 +82,10 @@ def tell(self, sentence):
         self.clauses.extend(conjuncts(to_cnf(sentence)))
 
     def ask_generator(self, query):
-        "Yield the empty substitution if KB implies query; else nothing."
+        "Return the empty substitution {} if KB entails query; else return False."
         if tt_entails(Expr('&', *self.clauses), query):
-            yield {}
+            yield {}  # Why use yield when you are not returning a generator?
```

Comment:
That's right, @chiragvartak -- ask_generator is designed for the question "tell me all the ways that this query could be true". Each way is a substitution. When there are no variables, it either yields the empty substitution once, saying there is one way to make the query true, or it doesn't yield anything at all, meaning there are no ways to make the query true. And I wanted to have the same interface for every kind of KB. I guess we could also add

     def ask_if_true(self, query): return self.ask(query) is not False